src/lang/eval/StateIPCClient.ml

```diff
@@ -0,0 +1,194 @@
+  | Ok res' -> pure res'
+
+let ipcclient_exn_wrapper thunk =
+  try
+    thunk()
+  with
+  | Unix.Unix_error (_, s1, s2) ->
+    fail0 ("StateIPCClient: Unix error: " ^ s1 ^ s2)
+  | _ -> fail0 "StateIPCClient: Unexpected error making JSON-RPC call"
+
+(* Send msg via socket s with a delimiting character "0xA". *)
+let send_delimited oc msg =
+  let msg' = msg ^ "\n" in
+  Out_channel.output_string oc msg';
+  Out_channel.flush oc
+
+let binary_rpc ~socket_addr (call: Rpc.call) : Rpc.response =
+  let socket = Unix.socket ~domain: Unix.PF_UNIX ~kind: Unix.SOCK_STREAM ~protocol:0 in
+  Unix.connect socket ~addr:(Unix.ADDR_UNIX socket_addr);
+  let (ic, oc) = Unix.in_channel_of_descr socket, Unix.out_channel_of_descr socket in
+  let msg_buf = Jsonrpc.string_of_call ~version: Jsonrpc.V2 call in
+  DebugMessage.plog (Printf.sprintf "Sending: %s\n" msg_buf);
+  (* Send data to the socket. *)
+  let _ = send_delimited oc msg_buf in
+  (* Get response. *)
+  let response = Caml.input_line ic in
+  Unix.close socket;
+  DebugMessage.plog (Printf.sprintf "Response: %s\n" response);
+  Jsonrpc.response_of_string response
+
+(* Encode a literal into bytes, opaque to the backend storage. *)
+let serialize_literal l = Bytes.of_string (PrettyPrinters.literal_to_jstring l)
+let deserialize_literal s tp =
+  try
+    pure @@ ContractState.jstring_to_literal s tp
+  with
+  | Invalid_json s ->
+    fail (s @ mk_error0 "StateIPCClient: Error deserializing literal fetched from IPC call")
+
+(* Map fields are serialized into Ipcmessage_types.MVal
+   Other fields are serialized using serialize_literal into bytes/string. *)
+let rec serialize_field value =
+  match value with
+  | Map (_, mlit) ->
+    let mpb = Caml.Hashtbl.fold (fun key value acc ->
+      let key' = Bytes.to_string (serialize_literal key) in
+      (* values can be Maps or non-map literals. Hence a recursive call. *)
+      let val' = serialize_field value in
+      (key', val') :: acc
+    ) mlit [] in
+    Ipcmessage_types.Mval({ Ipcmessage_types.m = mpb })
+    (* If there are maps _inside_ a non-map field, they are treated same
+     * as non-Map field values are not serialized as protobuf maps. *)
+  | _ -> Ipcmessage_types.Bval (serialize_literal value)
+
+(* Deserialize proto_scilla_val, given its type. *)
+let rec deserialize_value value tp =
+  match value with
+  | Ipcmessage_types.Bval s ->
+    deserialize_literal (Bytes.to_string s) tp
+  | Ipcmessage_types.Mval m ->
+    (match tp with
+    | MapType (kt, vt) ->
+      let mlit = Caml.Hashtbl.create (List.length m.m) in
+      let _ = iterM m.m ~f:(fun (k, v) ->
+        let%bind k' = deserialize_literal k kt in
+        let%bind v' = deserialize_value v vt in
+        Caml.Hashtbl.add mlit k' v';
+        pure ()
+      ) in
+      pure (Map ((kt, vt), mlit))
+    | _ -> fail0 "StateIPCClient: Type mismatch deserializing value. Unexpected protobuf map.")
```

Comment:
So such a map, since is serialized as a string (which is essentially our existing JSON string), it will be deserialized based on that, by our JSON parser.